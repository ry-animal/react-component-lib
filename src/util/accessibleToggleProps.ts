export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
