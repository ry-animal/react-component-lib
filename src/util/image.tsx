import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
