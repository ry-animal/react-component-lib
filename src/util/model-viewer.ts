/*
 * This file is a hacky way of allowing any other component to
 * use the <model-viewer /> tag just by doing:
 *   import "../../../util/model-viewer";
 *
 * The authors of the package do not expose it in a React-friendly
 * way, see: https://github.com/google/model-viewer/issues/1502
 *
 * The component documentation is at https://modelviewer.dev/docs/
 */

import "@google/model-viewer";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    export interface IntrinsicElements {
      "model-viewer": {
        src?: string;
        alt?: string;
        poster?: string;
        autoplay?: boolean;
        exposure?: number;
        "camera-controls"?: boolean;
        "environment-image"?: string;
        "seamless-poster"?: string;
        loading?: "auto" | "lazy" | "eager";
        reveal?: "auto" | "interaction" | "manual";
      } & React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement>,
        HTMLElement
      >;
    }
  }
}
