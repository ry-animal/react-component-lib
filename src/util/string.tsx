const DEFAULT_LOCALE = "en-US";

export const truncateEthAddress = (str: string): string => {
  if (str.length) {
    const numStartingCharacters = 6;
    const numTrailingCharacters = 4;
    return `${str.substring(0, numStartingCharacters)}...${str.substring(
      str.length - numTrailingCharacters,
      str.length
    )}`;
  }
  return "";
};

export const formatUSDPriceString = (value: number): string => {
  const minimumFractionDigits = 2;
  const maximumFractionDigits = 2;
  return `$${value.toLocaleString(DEFAULT_LOCALE, {
    minimumFractionDigits,
    maximumFractionDigits,
  })}`;
};

export const formatETHPriceString = (priceInEth: number): string => {
  const minimumFractionDigits = 0;
  const maximumFractionDigits = 8;
  const maxDigitsWithDecimal = 8;

  let formatted = `${priceInEth
    .toLocaleString(DEFAULT_LOCALE, {
      minimumFractionDigits,
      maximumFractionDigits,
    })
    .replace(/,/gu, "")
    .substring(0, maxDigitsWithDecimal)}`;

  if (formatted.endsWith(".")) {
    formatted = formatted.slice(0, -1);
  }
  return formatted;
};

const evaluateNumberLessThanOne = (num: number) => {
  if (num === 0) {
    return "0";
  }

  const minDecimalValue = 0.0001;
  if (num < minDecimalValue) {
    return minDecimalValue.toString();
  }
  // To round decimal to 4 digits
  const maxDigits = 10000;
  return `${Math.round(num * maxDigits) / maxDigits}`;
};

// Abbreviate Number https://stackoverflow.com/a/9462382

export const abbreviateNumber = (num: number): string => {
  const lookup = [
    { value: 1, symbol: "" },
    { value: 1e3, symbol: "k" },
    { value: 1e6, symbol: "M" },
    { value: 1e9, symbol: "B" },
    { value: 1e12, symbol: "T" },
    { value: 1e15, symbol: "P" },
    { value: 1e18, symbol: "E" },
  ];

  const rx = /\.0+$|(?<id>\.[0-9]*[1-9])0+$/u;
  const digits = 1;

  const item = lookup
    .slice()
    .reverse()
    .find((x) => num >= x.value);
  return item
    ? (num / item.value).toFixed(digits).replace(rx, "$1") + item.symbol
    : evaluateNumberLessThanOne(num);
};

export const getDomainFromRegex = (url: string) => {
  const matches = url.match(/:\/\/(?<id>.[^/]+)/u);
  // Extract domain (will be null if no match is found)
  return matches && matches[1];
};
