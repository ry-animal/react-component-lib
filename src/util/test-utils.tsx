import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
