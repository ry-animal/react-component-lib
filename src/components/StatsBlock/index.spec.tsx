import userEvent from "@testing-library/user-event";
import { StatsBlock } from ".";
import { render, screen } from "../../util/test-utils";

describe("<StatsBlock />", () => {
  const sampleProps = () => ({
    stats: [
      { label: "Floor Price", value: 0.00009, isETH: true, onClick: jest.fn() },
      { label: "Weekly Volume", value: 4024, isUSD: true },
      { label: "NFTs", value: 100000 },
      { label: "Owners", value: 45000 },
      { label: "For Sale", value: 10 },
    ],
  });

  describe("rendering", () => {
    test("should match snapshot", () => {
      expect(
        render(<StatsBlock {...sampleProps()} />).asFragment()
      ).toMatchSnapshot();
    });

    test("should render with dark theme styles", () => {
      expect(
        render(
          <StatsBlock {...sampleProps()} isDarkTheme={true} />
        ).asFragment()
      ).toMatchSnapshot();
    });
  });

  describe("event handling", () => {
    test("can be clicked", () => {
      const props = sampleProps();
      const [stat] = props.stats;
      const { onClick } = stat;
      render(<StatsBlock {...props} />);
      userEvent.click(screen.getByLabelText(stat.label));
      expect(onClick).toBeCalled();
    });
  });
});
