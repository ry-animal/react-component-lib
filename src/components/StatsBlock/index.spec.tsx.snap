// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StatsBlock /> rendering should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  border-radius: 8px;
  background-color: rgba(255,255,255,0.2);
  gap: 8px;
  padding: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 > * {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c3 {
  font-family: 'Poppins',sans-serif;
  border-radius: 4px;
  background-color: rgba(255,255,255,0.5);
  padding: 8px 12px;
  color: #000000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-width: 75px;
}

.c1 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  margin: 0;
  padding: 0;
  -webkit-transition: -webkit-transform 0.25s ease-out;
  -webkit-transition: transform 0.25s ease-out;
  transition: transform 0.25s ease-out;
}

.c1 .c2:hover,
.c1 .c2:focus {
  background-color: rgba(255,255,255,0.9);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 4px;
}

.c7 {
  text-align: center;
  font-weight: 500;
  font-size: 12px;
}

.c6 {
  font-weight: 400;
  font-size: 32px;
}

.c5 {
  width: 12px;
}

@supports (-webkit-backdrop-filter:blur(15px)) or (backdrop-filter:blur(15px)) {
  .c0 {
    -webkit-backdrop-filter: blur(15px);
    backdrop-filter: blur(15px);
  }
}

<div
    class="c0"
  >
    <button
      aria-label="Floor Price"
      class="c1"
    >
      <div
        class="c2 c3"
      >
        <div
          class="c4"
        >
          <img
            alt="ETH"
            class="c5"
            src="[object Object]"
          />
          <span
            class="c6"
          >
            &lt;0.0001
          </span>
        </div>
        <span
          class="c7"
        >
          Floor Price
        </span>
      </div>
    </button>
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <span
          class="c6"
        >
          $4,024.00
        </span>
      </div>
      <span
        class="c7"
      >
        Weekly Volume
      </span>
    </div>
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <span
          class="c6"
        >
          100k
        </span>
      </div>
      <span
        class="c7"
      >
        NFTs
      </span>
    </div>
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <span
          class="c6"
        >
          45k
        </span>
      </div>
      <span
        class="c7"
      >
        Owners
      </span>
    </div>
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <span
          class="c6"
        >
          10
        </span>
      </div>
      <span
        class="c7"
      >
        For Sale
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<StatsBlock /> rendering should render with dark theme styles 1`] = `
<DocumentFragment>
  .c0 {
  border-radius: 8px;
  background-color: rgba(255,255,255,0.2);
  gap: 8px;
  padding: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  background-color: rgba(255,255,255,0.2);
}

.c0 > * {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.c3 {
  font-family: 'Poppins',sans-serif;
  border-radius: 4px;
  background-color: rgba(255,255,255,0.5);
  padding: 8px 12px;
  color: #000000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-width: 75px;
  background-color: rgba(255,255,255,0.1);
  color: #ffffff;
}

.c1 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  margin: 0;
  padding: 0;
  -webkit-transition: -webkit-transform 0.25s ease-out;
  -webkit-transition: transform 0.25s ease-out;
  transition: transform 0.25s ease-out;
}

.c1 .c2:hover,
.c1 .c2:focus {
  background-color: rgba(255,255,255,0.3);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 4px;
}

.c7 {
  text-align: center;
  font-weight: 500;
  font-size: 12px;
}

.c6 {
  font-weight: 400;
  font-size: 32px;
}

.c5 {
  width: 12px;
}

@supports (-webkit-backdrop-filter:blur(15px)) or (backdrop-filter:blur(15px)) {
  .c0 {
    -webkit-backdrop-filter: blur(15px);
    backdrop-filter: blur(15px);
  }
}

<div
    class="c0"
  >
    <button
      aria-label="Floor Price"
      class="c1"
    >
      <div
        class="c2 c3"
      >
        <div
          class="c4"
        >
          <img
            alt="ETH"
            class="c5"
            src="[object Object]"
          />
          <span
            class="c6"
          >
            &lt;0.0001
          </span>
        </div>
        <span
          class="c7"
        >
          Floor Price
        </span>
      </div>
    </button>
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <span
          class="c6"
        >
          $4,024.00
        </span>
      </div>
      <span
        class="c7"
      >
        Weekly Volume
      </span>
    </div>
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <span
          class="c6"
        >
          100k
        </span>
      </div>
      <span
        class="c7"
      >
        NFTs
      </span>
    </div>
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <span
          class="c6"
        >
          45k
        </span>
      </div>
      <span
        class="c7"
      >
        Owners
      </span>
    </div>
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <span
          class="c6"
        >
          10
        </span>
      </div>
      <span
        class="c7"
      >
        For Sale
      </span>
    </div>
  </div>
</DocumentFragment>
`;
